Add tutorial: Where to start with bioimage analysis in Galaxy (#6603)
galaxyproject/training-material: topics/imaging/tutorials/where-to-start-bioimaging-galaxy/tutorial.md

Reviewer: I propose to replace: '> * **OME-Zarr:** The modern choice for "Big Data," cloud storage, and for image data in spatial transcriptomics. {% cite Moore2021 %}' by 'OME-Zarr: A cloud-optimized format for very large, multi-dimensional bioimaging datasets (including spatial transcriptomics), enabling scalable storage and efficient data access. {% cite Moore2021 %}'

Reviewer: Proposed revision for the paragraph to remove too close 'weights' instances and to emphasize that median filter remove impulsive noise typical of fluorescence microscopy: 'Denoising: Applies linear or non-linear filters to reduce noise while preserving meaningful structures. A Gaussian filter is a linear method that performs spatially weighted averaging using a kernel whose coefficients decrease smoothly with distance from the center, following a Gaussian (bell-shaped) distribution. It is well suited for reducing noise that approximates Additive White Gaussian Noise (AWGN). In contrast, a Median filter is a non-linear method that replaces each pixel value with the median of its local neighborhood. It is particularly effective for removing salt-and-pepper noise and is also appropriate for binary or label images because it preserves original intensity values. Compared to Gaussian smoothing, median filtering better maintains sharp edges—such as cell boundaries—while reducing impulsive noise often observed in fluorescence microscopy ({% cite Haase2022 %}).'

This tutorial acts as your **compass** to navigate the Galaxy imaging landscape, specifically as it applies to bioimage data pre-processing and analysis. You will establish the foundations necessary to transform raw, complex image data into structured tables of measurements, ready for downstream statistical analysis.

*To be add: image showing simple workflow, Raw image-Pixels, Numerical Grid for intensities, Data extraction, Final Data*

> <agenda-title></agenda-title>
>
> 1. TOC
> {:toc}
>
{: .agenda}

# 1. Know your data (the "digital anatomy" of an image)

As Pete Bankhead points out in *[Introduction to Bioimage Analysis](https://bioimagebook.github.io/index.html)* ({% cite Bankhead2022 %}), an image is not just a picture, it is a collection of measurements. Therefore, before starting any analysis, you must understand the underlying structure of your data.

> To the human eye, an image is a visual representation of a biological sample. However, to a computer, it is a **numerical matrix** ({% cite Sedgewick2010 %}). Every point in that grid—the **pixel**—is a data point representing the number of photons or the signal intensity detected at that specific coordinate. Understanding your "digital anatomy" means knowing exactly how those numbers were recorded, how they are spaced in 3D space, and what the intensity limits are. 

Author: You are correct! I will introduce the term tensor.

*Add Pete B's Image from his book*

If you don't understand the numbers behind the colors, you risk performing "Image processing" (simply making a pretty picture) rather than "Image analysis" (extracting scientific truth). Let's unpack the data behind an image step by step. 

## Pixels and voxels

> An image is a grid of **pixels** (2D) or **voxels** (3D). Think of an image as a vast mosaic where every tile is a "picture element." In 2D, these tiles are flat, but in 3D imaging, they have depth and are called **voxels** ("volumetric elements") ({% cite Pawley2006 %}). Each one is like a small bucket that has captured a specific amount of signal (usually light), which the computer records as a single number.

Reviewer 2: Just some typesetting nitpicking: (it's kind of bad practice to have concatenations of parentheses like `)(` because it's not clear then what the second parentheses refers to, the former parentheses or the text before that); and since `Pawley2006` isn't the one who introduced voxels, it's good practice to prefix it with an `e.g.,` to indicate that it is just an example and not an attribution:

```suggestion
An image is a grid of **pixels** (2D) or **voxels** (3D). Think of an image as a vast mosaic where every tile is a "picture element." In 2D, these tiles are flat, but in 3D imaging, they have depth and are called **voxels** ("volumetric elements", e.g., {% cite Pawley2006 %}). Each one is like a small bucket that has captured a specific amount of signal (usually light), which the computer records as a single number.
```

> An image is a grid of **pixels** (2D) or **voxels** (3D). Think of an image as a vast mosaic where every tile is a "picture element." In 2D, these tiles are flat, but in 3D imaging, they have depth and are called **voxels** ("volumetric elements") ({% cite Pawley2006 %}). Each one is like a small bucket that has captured a specific amount of signal (usually light), which the computer records as a single number.

Reviewer 2: The imaging device isn't necessarily a "computer" in the classic sense, is it?
```suggestion
An image is a grid of **pixels** (2D) or **voxels** (3D). Think of an image as a vast mosaic where every tile is a "picture element." In 2D, these tiles are flat, but in 3D imaging, they have depth and are called **voxels** ("volumetric elements") ({% cite Pawley2006 %}). Each one is like a small bucket that has captured a specific amount of signal (usually light), which the imaging device records as a single number.
```

To make these numbers useful for science, we must define their **range** and their **size**.

> ### Bit depth (the range)

Reviewer 2: In this section, you are referring to _integer representations_ of intensity values, which is kind of a natural representation if you see an imaging device as a photon counter. But there are also floating point representations, which also have different bit depths, but those have different limits. Somehow this should also be mentioned here.

This defines the precision of your "buckets" by setting the range of possible intensity values. 

* **8-bit:** $2^{8} = 256$ levels ($0–255$). While this looks fine to our eyes, it is often too "coarse" for sensitive mathematical analysis.

> * **16-bit:** $2^{16} = 65,536$ levels ($0–65,535$). This is the **scientific gold standard** because it allows you to detect tiny differences in brightness that would be lost or rounded away in 8-bit ({% cite Haase2022 %}).

Reviewer 2: ```suggestion
* **16-bit:** $2^{16} = 65,536$ levels ($0–65,535$). This is the **scientific gold standard** because it allows you to detect subtle differences in image intensities that would be lost (e.g., due to rounding) when using an 8-bit representation (e.g., {% cite Haase2022 %}).
```

### Spatial calibration (the size)
By itself, a pixel has no physical size; it is just a unit of storage. **Spatial Calibration** is the metadata that links these digital units to physical reality (e.g., $1 \text{ pixel} = 0.25 \mu m$). Without this "secret sauce," you can count objects, but you cannot accurately measure how big they are, how fast they move, or their concentration ({% cite Linkert2010 %}, {% cite Haase2022 %}). 

This calibration is usually stored in the image header. If you lose this metadata during a file conversion (e.g., saving as a standard .jpg), your analysis will only be able to provide results in "pixels," which have no biological meaning in a publication.

> <tip-title> Avoid Saturation </tip-title> 
> If you see a "0," the sensor detected nothing. If you see the maximum value ($255$ or $65,535$), your sensor was overwhelmed, this is called **saturation**. Saturated pixels are often "clipped," meaning the true biological signal was higher than what the camera could record ({% cite Pawley2006 %}). This data is lost forever and cannot be accurately quantified. 
>
{: .tip}

## The 5 dimensions (5D)

> In everyday photography, we usually deal with 2D color images. However, in the life sciences, we can capture **hyperstacks**: multi-dimensional data structures that represent a biological event across space, spectrum, and time. Most scientific images follow the $(X, Y, Z, C, T)$ convention ({% cite Goldberg2005 %}):

Reviewer 2: "event" seems like a misplaced term here. I think it refers to something that happens in time, doesn't it? Maybe:

```suggestion
In everyday photography, we usually deal with 2D color images. However, in the life sciences, we can capture **hyperstacks**: multi-dimensional data structures that represent a biological sample across space, spectrum, and time. Most scientific images follow the $(X, Y, Z, C, T)$ convention ({% cite Goldberg2005 %}):
```

Also, I see that you describe X, Y, and so on, below, but it'd be good to briefly mention what the XYZCT convention is, before going into the definitions of the respective letters.

* **X & Y (Spatial):** The width and height of your image.
* **Z (Depth):** Multiple optical sections or "slices" taken at different focal planes to reconstruct a 3D volume.
* **C (Channel):** Different "colors" or fluorescent probes (e.g., DAPI for nuclei, GFP for proteins).
* **T (Time):** Separate frames captured over a duration (time-lapse), allowing you to track movement or dynamic changes.

However, it is important to note that while we describe them in this order, different microscope vendors store them in different internal sequences (e.g., $TCZXY$). Using standard metadata readers is essential to ensure Galaxy doesn't accidentally interpret a Z-slice as a Time-point ({% cite Linkert2010 %}).

> <tip-title> Respect the channels </tip-title> 
> While we often look at "merged" RGB images for presentations, you should **always perform quantification on the raw, individual channels.** Merging or converting to RGB often involves data compression, bit-depth reduction, or "clipping" that distorts the underlying intensity measurements ({% cite Cromey2010 %}). 
> 
{: .tip}

> <question-title> Identify the Dimensions </question-title>
>

> > If you have a time-lapse experiment where you filmed 3 different fluorescent markers across 10 focal planes (Z-slices) every minute for an hour, how many total 2D images (planes) are in your file?

Reviewer 2: I think "filming" is rather a term common when describing movies…

```suggestion
> If you have a time-lapse experiment where you recorded 3 different fluorescent markers across 10 focal planes (Z-slices) every minute for an hour, how many total 2D images (planes) are in your file?
```

*Add Pete B's Image from his book*

If you don't understand the numbers behind the colors, you risk performing "Image processing" (simply making a pretty picture) rather than "Image analysis" (extracting scientific truth). Let's unpack the data behind an image step by step. 

## Pixels and voxels

An image is a grid of **pixels** (2D) or **voxels** (3D). Think of an image as a vast mosaic where every tile is a "picture element." In 2D, these tiles are flat, but in 3D imaging, they have depth and are called **voxels** ("volumetric elements", e.g., {% cite Pawley2006 %}). Each one is like a small bucket that has captured a specific amount of signal (usually light), which the computer records as a single number.

> 

Reviewer 2: Add this here so it will be clearer below what "representations" are about:

```suggestion
Image intensities represent the measurements of the imaging system.

```

Reviewer 3: breaks the flow a little bit for me

To make these numbers useful for science, we must define their **range** and their **spatial extent** (i.e. the *resolution* of the image, like pixels or voxels per real-world unit).

### Bit depth (the range and precision limits)

> Bit depth sets the **maximum theoretical precision** of your measurements by determining the range of possible intensity values and the smallest distinguishable difference between them. However, the **actual achievable precision** depends on many factors in your imaging system: sensor noise, photon statistics, optical aberrations, and sample preparation quality. In other words, bit depth imposes an upper bound—the real precision may be lower due to experimental limitations (e.g., {% cite Pawley2006 %}).

Reviewer 2: I'd avoid "measurements" because it might be confusing (e.g, readouts from analyses).

```suggestion
The bit depth sets the **maximum possible precision** to store image intensities by determining the range of possible intensity values and the smallest distinguishable difference between them. However, the **actual precision** also depends on many factors in your imaging system: sensor noise, photon statistics, optical aberrations, and sample preparation quality. In other words, the bit depth imposes an upper bound—the real precision may be lower due to experimental limitations (e.g., {% cite Pawley2006 %}).
```

or maybe:

```suggestion
The bit depth corresponds to an **upper limit of the precision** when storing image intensities by determining the range of possible intensity values and the smallest distinguishable difference between them. The **actual precision** may also be lower and depends on many factors in your imaging system, like sensor noise, photon statistics, optical aberrations, and sample preparation quality (e.g., {% cite Pawley2006 %}).
```

* **16-bit:** $2^{16} = 65,536$ discrete levels ($0–65,535$). This is the **scientific gold standard for acquisition** because it allows you to detect subtle differences in image intensities that would be lost (e.g., due to rounding) when using an 8-bit representation (e.g., {% cite Haase2022 %}).

#### Floating-point representations

After image processing operations (e.g., background subtraction, normalization, deconvolution), intensity values often become non-integer or fall outside the original acquisition range. Floating-point formats accommodate this:

* **16-bit float (half precision):** Can represent a range of approximately $6.1 \times 10^{-5}$ to $65,504$ with decimal precision. This format is increasingly used in machine learning and GPU-accelerated image processing because it saves memory while providing sufficient precision for many applications.
* **32-bit float (single precision):** Can represent an enormous range (~$10^{-38}$ to ~$10^{38}$) with ~7 significant decimal digits of precision. This is the **most common format for image processing** as it balances precision, range, and computational efficiency.

> * **64-bit float (double precision):** Offers even greater precision (~15-16 significant decimal digits) and range (~$10^{-308}$ to ~$10^{308}$), useful for iterative algorithms or when accumulating many operations where small errors could compound.

Reviewer 2: This reads as if floating point representations can only represent positive values, which is not the case. Let me break down what the limits that you mention actually are about. So you mention two limits for each floating point representation, a lower limit and an upper limit.

The _upper limits_ are the upper limits of representable *absolute values*: Each floating point representation is capable to represent values in $[-f; +f]$ (e.g., $f \approx 10^{38}$ for float-32). Each floating point representation is also capable to exactly represent 0. The _lower limits_ are the lower limits of the *non-zero absolute values* that can be represented (e.g., float-32 cannot represent values $f$ with $|f| < 10^{-38}$ and $f \neq 0$).

In practice, I think, the lower limits are not that important. For the sake of complexity, I'd only mention the upper limits here, or maybe describe them in a less formal way.

This calibration information is usually stored in the image header. If you lose this metadata during a file conversion (e.g., saving as a standard .jpg), your analysis will only be able to provide results in "pixels," which have no biological meaning in a publication.

> <tip-title> Avoid saturation </tip-title> 
> If you see an image intensity value of "0", the sensor may have detected nothing, or it could represent true absence of signal—context matters. If you see the maximum value (e.g., $255$ or $65,535$), your sensor was overwhelmed. These phenomena are called **saturation** (under- and oversaturation, respectively). Saturated pixels are often "clipped," meaning the true biological signal was lower or higher than what the camera could record (e.g., {% cite Pawley2006 %}). This data is lost forever and cannot be accurately quantified. 
>
{: .tip}

> ## The 5 dimensions (5D)

Reviewer 2: I think this section should go right before `Pixels and voxels`.

> 
{: .tip}

> <question-title> Identify the dimensions </question-title>
>
> If you have a time-lapse experiment where you recorded 3 different fluorescent markers across 10 focal planes (Z-slices) every minute for an hour, how many total 2D images (planes) are in your file?
>
> > <solution-title></solution-title>
> >
> > You would have $3 \text{ (Channels)} \times 10 \text{ (Z-slices)} \times 60 \text{ (Time points)} = 1,800$ total 2D planes. Each individual plane is an $X \times Y$ image. Galaxy tools need to know the correct dimensional order (e.g., $XYZCT$ vs. $XYCZT$ vs. $TZCXY$) to ensure they correctly interpret which dimension is which—otherwise, you might accidentally measure intensity changes across Z-depth when you meant to measure changes over time!
> >
> {: .solution} 
{: .question}

## Bit depth: why it matters for science

> A common misconception is that if two images look identical on a desktop monitor, they contain the same data. However, our eyes and screens are usually limited to 8-bit displays, while your microscope sensor is far more sensitive. 

Reviewer 2: ```suggestion
A common misconception is that if two images look identical on a desktop monitor, they contain the same data. However, our eyes are used to computer screens that are usually limited to 8-bit, while your microscope sensor is far more sensitive. 
```

When you perform certain pre-processing tasks, such as subtracting the image background or contrast enhancement, you are essentially **stretching the histogram** of the image data.

* In **8-bit**, with only 256 possible values, stretching creates more pronounced "gaps" in your histogram (quantization artifacts), potentially making your data distribution appear discontinuous.
* In **16-bit integer**, with 65,536 possible values, the same stretching operation creates smaller relative gaps. While quantization artifacts still occur, they are far less severe and less likely to impact downstream quantitative analysis ({% cite Cromey2010 %}).

> The key difference is not whether artifacts appear, but their **magnitude relative to your data range**. Think of it this way: spreading 256 values across a wider range creates larger "jumps" between adjacent intensity levels than spreading 65,536 values across the same range.

Reviewer 2: ```suggestion
The key difference is not whether artifacts appear, but their **magnitude in relation to your data range**. Think of it this way: spreading 256 values across a wider range creates larger "holes" between adjacent intensity levels than spreading 65,536 values across the same range.
```

* **The Basics:** [Introduction to Image Analysis]({% link topics/imaging/tutorials/imaging-introduction/tutorial.md %}) – A deeper dive into the fundamental concepts of digital images.

* **Fluorescence & Screening:** [HeLa Cell Screen Analysis]({% link topics/imaging/tutorials/hela-screen-analysis/tutorial.md %}) – Learn to process high-throughput screens using classical techniques.

* **Time-Lapse & Events:** [Detection of MitoFlashes]({% link topics/imaging/tutorials/detection-of-mitoflashes/tutorial.md %}) – Track transient biological events over time.

* **Classical Segmentation:** [Voronoi-based Segmentation]({% link topics/imaging/tutorials/voronoi-segmentation/tutorial.md %}) – A powerful approach for partitioning cells when they are touching but have clear centers.

* **Advanced AI:** [Imaging using Bioimage Zoo Models]({% link topics/imaging/tutorials/process-image-bioimageio/tutorial.md %}) – Use pre-trained deep learning models for complex segmentation tasks.

* **Automated Pipelines:** [CellProfiler in Galaxy]({% link topics/imaging/tutorials/tutorial-CP/tutorial.md %}) – Master the use of CellProfiler modules to build end-to-end automated workflows.

> * **Tracking:** [Object Tracking using CellProfiler]({% link topics/imaging/tutorials/object-tracking-using-cell-profiler/tutorial.md %}) – Move from static images to following individual objects through time and space.

Reviewer 2: Some cosmetics: `{% icon level %}` looks like 🎓

See [this](https://training.galaxyproject.org/training-material/topics/single-cell/tutorials/scrna-case_alevin/tutorial.html#warning-for-the-bench-scientists-and-biologists) for an example

```suggestion
* **The Basics:** {% icon level %} [Introduction to Image Analysis]({% link topics/imaging/tutorials/imaging-introduction/tutorial.md %}) – A deeper dive into the fundamental concepts of digital images.

* **Fluorescence & Screening:** {% icon level %} [HeLa Cell Screen Analysis]({% link topics/imaging/tutorials/hela-screen-analysis/tutorial.md %}) – Learn to process high-throughput screens using classical techniques.

* **Time-Lapse & Events:** {% icon level %} [Detection of MitoFlashes]({% link topics/imaging/tutorials/detection-of-mitoflashes/tutorial.md %}) – Track transient biological events over time.

* **Classical Segmentation:** {% icon level %} [Voronoi-based Segmentation]({% link topics/imaging/tutorials/voronoi-segmentation/tutorial.md %}) – A powerful approach for partitioning cells when they are touching but have clear centers.

* **Advanced AI:** {% icon level %} [Imaging using Bioimage Zoo Models]({% link topics/imaging/tutorials/process-image-bioimageio/tutorial.md %}) – Use pre-trained deep learning models for complex segmentation tasks.

* **Automated Pipelines:** {% icon level %} [CellProfiler in Galaxy]({% link topics/imaging/tutorials/tutorial-CP/tutorial.md %}) – Master the use of CellProfiler modules to build end-to-end automated workflows.

* **Tracking:** {% icon level %} [Object Tracking using CellProfiler]({% link topics/imaging/tutorials/object-tracking-using-cell-profiler/tutorial.md %}) – Move from static images to following individual objects through time and space.
```

---
layout: tutorial_hands_on

title: "Where to start with bioimage analysis in Galaxy"
level: Introductory
subtopic: introduction
questions:

>   - I have images from my experimental data. Can I process them in Galaxy? Where do I begin?

Reviewer 3: ```suggestion
  - Can I process images from my experiments in Galaxy? Where do I begin?
```

* **64-bit float (double precision):** Offers even greater precision (~15-16 significant decimal digits) and range (~$10^{-308}$ to ~$10^{308}$), useful for iterative algorithms or when accumulating many operations where small errors could compound.

Note that floating-point images have **different limits** than integer images—instead of a fixed maximum like 255 or 65,535, the "range" depends on your actual data and processing steps. Many analysis tools automatically convert to floating-point internally to preserve accuracy during calculations.

### Spatial calibration (the size)

Image pixels (or voxels) have no _intrinsic_ physical size; they are just units of storage and representation. **Spatial calibration** is the metadata that links these digital units to physical reality (e.g., $1 \text{ pixel} = 0.25 \mu m$). Without this "secret sauce," you can count objects, but you cannot accurately measure how big they are, how fast they move, or their concentration (e.g., {% cite Linkert2010 %}, {% cite Haase2022 %}). 

> Metadata like the calibration information is usually stored in the image header. If you lose this metadata during a file conversion (e.g., saving as a standard .jpg), your analysis will only be able to provide results in "pixels," which have no biological meaning in a publication.

Reviewer 3: not just for publication purposes, right?

{: .tip} 

In Galaxy, you can start the pre-processing stage with tools like **{% tool [Apply standard image filter](toolshed.g2.bx.psu.edu/repos/imgteam/2d_simple_filter/ip_filter_standard/1.16.3+galaxy1) %}**.

> <hands-on-title> Filtering Noise </hands-on-title>
>
> 1. {% tool [Apply standard image filter](toolshed.g2.bx.psu.edu/repos/imgteam/2d_simple_filter/ip_filter_standard/1.16.3+galaxy1) %} with the following parameters:
>    - {% icon param-select %} *"Type of image data to process"*: `2-D image data (or series thereof)`
>        - {% icon param-file %} *"Input image (2-D)"*: `your_uploaded_image.tif`
>        - {% icon param-select %} *"Filter type"*: `Median`
>            - {% icon param-text %} *"Size"*: `3`
>
> > <comment-title>Choosing the right filter</comment-title>
> > Try running this tool twice: once with a **Median** filter (using a small *Size* like 3) and once with a **Gaussian** filter (adjusting *Sigma*). Zoom in on the edges of your objects. You will notice that the Median filter preserves boundaries much better, while the Gaussian filter blurs them.
> {: .comment}
>
{: .hands_on}

> ## Stage B: Segmentation (Defining objects)

Reviewer 3: ```suggestion
## Stage B: Segmentation (defining objects)
```

>
> 2. **Review the output:** You should now have a binary image where your biological objects are represented as white pixels (255) on a black background (0).
>
{: .hands_on}

> <comment-title>Masks vs. Label Images</comment-title>
> It is important to distinguish between these two types of images in Galaxy:
>
> * **Binary Mask:** This is what you created with the **Threshold** tool. Every pixel is either "Background" (0) or "Foreground" (255). It tells the computer *where* the objects are, but it doesn't know that one cell is different from the one next to it.
> * **Label Image:** Tools like **Label objects** (Connected Component Analysis) take a mask and assign a unique integer to every separate "blob." The first cell’s pixels all become `1`, the second cell `2`, and so on.
>
{: .comment}

## Stage B.2: The Region of Interest (ROI)

Once you have segmented your image, you have created **Regions of Interest (ROIs)**. This is a central concept in bioimaging. An ROI is a spatial selection that tells the software: "Ignore the background; only calculate values for these specific coordinates."

In Galaxy, ROIs can take three forms depending on the tool you use:

> * **Bounding Boxes:** Simple rectangles used to crop an image or define a search area. In Galaxy, this is only possible to perform this step using interactive tools like Qupath.

Reviewer 3: ```suggestion
* **Bounding Boxes:** Simple rectangles used to crop an image or define a search area. In Galaxy, this is only possible to perform this step using interactive tools like QuPath.
```

> > 2. **Volume:** With 500 images, you need the batch-processing power of **CellProfiler** to stay efficient.
> > 
> {: .solution}
{: .question}

> <question-title> Scenario 2: The 3D Volume </question-title>
> You have a single 3D Z-stack of a zebrafish embryo with one fluorescent marker. What is your path?
> > <solution-title></solution-title>
> 
> > 1. **Microscopy?** Yes.
> > 2. **Color or B&W?** B&W (Single-channel fluorescence).
> > 3. **Dimensionality?** 3D volume.
> > **Result:** You are on the **Segmentation** path. Use **Cellpose** for automated results or **QuPath** interactively to test different StarDist models.
> > 
> {: .solution}
{: .question}
 

# 6. Common pitfalls to avoid

Even with the best tools, it is easy to accidentally "break" your data before you even start measuring. In bioimage analysis, these errors are often called **artifacts**.

## 1. The "JPG" trap
**Never use JPEG for science.** JPEGs use "lossy" compression, meaning the computer slightly changes pixel intensities to save space. This creates "blocky" artifacts that ruin your ability to measure protein concentration or fine textures.

> * **The fix:** Always use **TIFF**, **OME-TIFF**, or **PNG**. These are "lossless" formats that preserve every photon counted by the camera.

Reviewer 2: Maybe add a comment for when PNG is suitable (2-D single-channel or RGB images that are not too large, e.g., 2-D label images, 2-D segmentation masks, images for visualization purposes).

So the bottom line is, IMO, there are cases when PNG really is superior to TIFF, because of the file size that might be smaller by several magnitudes. But of course you have to be careful to check whether you have such a case.

Metadata like the calibration information is usually stored in the image header. If you lose this metadata during a file conversion (e.g., saving as a standard .jpg), your analysis will only be able to provide quantitative results in "pixels," which may have no biological meaning.

> <tip-title> Avoid saturation </tip-title> 
> If you see an image intensity value of "0", the sensor may have detected nothing, or it could represent true absence of signal—context matters. If you see the maximum value (e.g., $255$ or $65,535$), your sensor was overwhelmed. These phenomena are called **saturation** (under- and oversaturation, respectively). Saturated pixels are often "clipped," meaning the true biological signal was lower or higher than what the camera could record (e.g., {% cite Pawley2006 %}). This data is lost forever and cannot be accurately quantified. 
>
{: .tip}

# 2. How to get your images into Galaxy

> Galaxy is built to handle the complexity of biological data. However, microscopy images often come in "vendor-specific" formats. Your entry point into Galaxy depends on how your data was saved:

Reviewer 2: It's also not the same, strictly speaking. Proprietary is the opposite of open, right? While vendor-specific formats can still be open. Depends on what you want to say.

* **Standard Formats (.tiff, .png):** Use the standard Galaxy **Upload** tool.
* **Open Formats:** OME-tiff that includes metadata and OME-Zarr for spatial transcriptomics.
* **Proprietary Formats (.czi, .nd2, .lif):** These formats "wrap" image data and metadata together. While you can often export TIFFs from your microscope software, using the **{% tool [Convert image format with Bioformats](toolshed.g2.bx.psu.edu/repos/imgteam/bfconvert/ip_convertimage/6.7.0+galaxy3) %}** tool allows Galaxy to "unlock" and standardize the metadata hidden inside these files ({% cite Moore2021 %}).
* **OMERO Integration:** If your institution uses an **OMERO server**, you can import images directly via the **Remote Files** section in the upload tool. You can access OMERO using galaxy. See [Overview of the Galaxy OMERO-suite]({% link topics/imaging/tutorials/omero-suite/tutorial.md %})

> ## Why use the Bio-Formats tool suite?

Reviewer 4: Maybe also add a link to the bioformat page:
https://bio-formats.readthedocs.io/en/v8.4.0/

> ## Why use the Bio-Formats tool suite?

Reviewer 4: ```suggestion
## Why use the Bio-Formats tool?
```

>
> While there are many options, two open community-endorsed formats are increasingly dominant in bioimaging:
> * **OME-TIFF:** Best for "classic" microscopy (2D/3D stacks) and maximum compatibility with classical software like ImageJ/Fiji or QuPath.
> * **OME-Zarr:** The modern choice for "Big Data," cloud storage, and for image data in spatial transcriptomics. {% cite Moore2021 %}
> 
> Both are superior to proprietary formats because they make it possible to preserve your metadata attached to your pixel data throughout the entire Galaxy workflow.
>
{: .tip}

> <question-title> Metadata testing </question-title>
>
> Why we should avoid bioimages to be "Save as JPEG" from your microscope or computer?
>
> > <solution-title></solution-title>
> >
> > 1. **Data Integrity:** JPEGs use "lossy" compression, which changes pixel values to save space. This approach literally modifies your scientific data.
> > 2. **Metadata Retention:** A standard JPEG or simple TIFF often "strips" the metadata. You would lose the information about the physical size a pixel represents, making it impossible to calculate the size of your objects later.
> >
> {: .solution}
{: .question}

> # 3. Before you begin: diagnose your data and define your goals

Reviewer 4: ```suggestion
# 3. Before you begin: inspect your data and define your goals
```

>
> Option 2: Cellpose-SAM (Next Gen)
> 1. {% tool [Run generalist cell and nucleus segmentation with Cellpose-SAM](toolshed.g2.bx.psu.edu/repos/bgruening/cp_sam/cellposesam/4.0.8+galaxy0) %} with the following parameters:
>    - {% icon param-file %} *"Choose the image file for segmentation"*: `your_preprocessed_image.tif`
>    - {% icon param-check %} *"Save masks as tiff?"*: `Yes` (Required for quantification later)
>    - {% icon param-check %} *"Save RGB outline images"*: `Yes` (Useful for Stage E: Validation)
>    - In **Advanced Options**:
>        - {% icon param-text %} *"Cell or nuclei diameter in pixels"*: `Enter approximate diameter`
>        - {% icon param-check %} *"Whether to use transformer backbone"*: `Yes` (to utilize the Cellpose3 transformer model)
>
> > <comment-title>Multi-channel Images</comment-title>
> > Cellpose tools in Galaxy typically expect single-channel 2D images. If you have a multi-channel image, you must first use the **Split image along axes** tool to extract the specific channel (e.g., the DAPI channel for nuclei) that you wish to segment.
> {: .comment}
{: .hands_on}

> ### B. Interactive tools (Visual exploration)

Reviewer 4: Napari? :)

## Stage D: Quantification (Extracting numbers)
Now that you have a clean mask, Galaxy "overlays" it back onto your original raw image. Since the computer knows exactly which pixels belong to which cell, it can now do the math:
* **Intensity:** "How much protein (fluorescence) is inside this specific mask?"
* **Morphology:** "How large, round, or elongated is this nucleus?"
* **Spatial distribution:** "How far is the nucleus from the cell membrane?"

## Stage E: Validation (The sanity check)

**Never trust an algorithm blindly.** The final step is a visual inspection to ensure your numerical outputs represent biological reality rather than processing artifacts ({% cite Bankhead2022 %}). 

* **{% tool [Overlay outlines](toolshed.g2.bx.psu.edu/repos/bgruening/cp_overlay_outlines/cp_overlay_outlines/3.1.9+galaxy1) %}**: This tool takes your **original image** and draws the boundaries of your **detected objects** (from your Label Image) on top of it as colored lines. It is the gold standard for visual validation, allowing you to confirm that the computer's segmentation accurately followed the biological edges of your cells.

> * **{% tool [Overlay images](toolshed.g2.bx.psu.edu/repos/imgteam/overlay_images/ip_overlay_images/0.0.5) %}**: A versatile tool for visualizing colocalization or blending two different channels to see where biological signals overlap.

Reviewer 2: Also capable of overlaying segmentation outlines.

 - **What comparison will you make?** (Differences between conditions, changes across time, or descriptive characterization?)

Depending on your answers, your starting path in Galaxy will change.

# 4. The lifecycle of an analysis pipeline

A typical project in Galaxy is not a single click, but a sequence of logical steps. Think of it as a factory assembly line: you start with raw materials (pixels) and move through various stations until you have a finished product (a table of measurements). Let's examine these stages together.

## Stage A: Pre-processing (cleaning)
Raw images are rarely perfect. They often contain electronic noise from the camera, uneven illumination from the microscope, or staining artifacts. Pre-processing prepares your digital image for analysis by enhancing the structures you want to identify and suppressing unwanted features like noise or artifacts ({% cite Bankhead2022 %}).

* **Background subtraction:** Removes the "haze" or background fluorescence caused by out-of-focus light. This is crucial for accurate intensity measurements later on.

> * **Denoising:** Applies linear or non-linear filters to reduce noise while preserving meaningful structures. A **Gaussian filter** is a linear method that performs spatially weighted averaging using a kernel whose coefficients decrease smoothly with distance from the center, following a Gaussian (bell-shaped) distribution. It is well suited for reducing noise that approximates [Additive White Gaussian Noise (AWGN)](https://en.wikipedia.org/wiki/Additive_white_Gaussian_noise). In contrast, a **Median filter** is a non-linear method that replaces each pixel value with the median of its local neighborhood. It is particularly effective for removing [salt-and-pepper noise](https://en.wikipedia.org/wiki/Salt-and-pepper_noise) and is also appropriate for binary or label images because it preserves original intensity values. Compared to Gaussian smoothing, median filtering better maintains sharp edges such as cell boundaries (e.g., {% cite Haase2022 %}).

Reviewer 2: Thanks @arrmunoz! Since salt-and-pepper noise is a sub-class of impulse noise, I suggest to phrase like:
```suggestion
* **Denoising:** Applies linear or non-linear filters to reduce noise while preserving meaningful structures. A **Gaussian filter** is a linear method that performs spatially weighted averaging using a kernel whose coefficients decrease smoothly with distance from the center, following a Gaussian (bell-shaped) distribution. It is well suited for reducing noise that approximates [additive white Gaussian noise (AWGN)](https://en.wikipedia.org/wiki/Additive_white_Gaussian_noise). In contrast, a **Median filter** is a non-linear method that replaces each pixel value with the median of its local neighborhood. It is particularly effective for removing impulse noise such as [salt-and-pepper noise](https://en.wikipedia.org/wiki/Salt-and-pepper_noise) and can also help mitigate impulse-like components of [Poisson (shot) noise](https://en.wikipedia.org/wiki/Shot_noise). It is also appropriate for binary or label images because it preserves original intensity values. Compared to Gaussian smoothing, median filtering better maintains sharp edges such as cell boundaries (e.g., {% cite Haase2022 %}).
```
Now we have it all included, I think?

I also made `Additive White Gaussian Noise` lowercase to be consistent with `salt-and-pepper noise`.

Reviewer: looks good. Thank you!

> * **Denoising:** Applies linear or non-linear filters to reduce noise while preserving meaningful structures. A **Gaussian filter** is a linear method that performs spatially weighted averaging using a kernel whose coefficients decrease smoothly with distance from the center, following a Gaussian (bell-shaped) distribution. It is well suited for reducing noise that approximates [Additive White Gaussian Noise (AWGN)](https://en.wikipedia.org/wiki/Additive_white_Gaussian_noise). In contrast, a **Median filter** is a non-linear method that replaces each pixel value with the median of its local neighborhood. It is particularly effective for removing [salt-and-pepper noise](https://en.wikipedia.org/wiki/Salt-and-pepper_noise) and is also appropriate for binary or label images because it preserves original intensity values. Compared to Gaussian smoothing, median filtering better maintains sharp edges—such as cell boundaries—while reducing impulsive noise often observed in fluorescence microscopy ({% cite Haase2022 %}).

Author: Could you please make a suggestion directly? I think the tutorial should be merge soon. Thank you!

>
> * **Binary Mask:** This is what you created with the **Threshold** tool. Every pixel is either "Background" (0) or "Foreground" (255). It tells the computer *where* the objects are, but it doesn't know that one cell is different from the one next to it.
> * **Label Image:** Tools like **Label objects** (Connected Component Analysis) take a mask and assign a unique integer to every separate "blob." The first cell’s pixels all become `1`, the second cell `2`, and so on.
>
{: .comment}

### Stage B.1: The Region of Interest (ROI)

Once you have segmented your image, you have created **Regions of Interest (ROIs)**. This is a central concept in bioimaging. An ROI is a spatial selection that tells the software: "Ignore the background; only calculate values for these specific coordinates."

In Galaxy, ROIs can take three forms depending on the tool you use:
* **Bounding Boxes:** Simple rectangles used to crop an image or define a search area. In Galaxy, this is only possible to perform this step using interactive tools like QuPath.
* **Binary Masks:** A "stencil" where the ROI is white and everything else is black.

> * **Label Images:** A "smart map" where every individual ROI has its own unique integer ID ({% cite Tosi2021 %}).

Reviewer 2: ```suggestion
* **Label Images:** A "smart map" where every individual ROI has its own unique integer ID (e.g., {% cite Tosi2021 %}).
```

* **The fix:** Always store quantitative microscopy data in **TIFF** or **OME-TIFF**. These formats preserve the original pixel values and metadata.

* **When is PNG OK?** PNG is lossless and suitable for small 2-D single-channel or RGB images where intensities are not used for quantitative measurements. For example: 2-D label images, segmentation masks, or visualization outputs. In these cases it can reduce file size by orders of magnitude compared to TIFF. However, it should not be used for raw or measurement-grade microscopy data.

## 2. The "merged image" mistake

Analyzing a "Merge" (RGB) image is risky because the intensities of different channels (like DAPI and GFP) are mathematically blended into a single color value.

* **The fix:** Always **Split Channels** in Galaxy. For this purpose you may use the {% tool [Split image along axes with NumPy](toolshed.g2.bx.psu.edu/repos/imgteam/split_image/ip_split_image/2.3.5+galaxy0) %} tool. Measure your DAPI (nuclei) and GFP (protein) separately to ensure scientific accuracy.

## 3. Ignoring saturation

> If your image is too bright, you might hit the camera sensor's limit ($255$ for 8-bit or $65,535$ for 16-bit). This is called **Clipping**. 

Reviewer 2: ```suggestion
If your image is too bright, you might hit the camera sensor's limit ($$255$$ for 8-bit or $$65,535$$ for 16-bit). This is called **Clipping**. 
```

* **Standard Formats (.tiff, .png):** Use the standard Galaxy **Upload** tool.

> * **Open Formats:** OME-TIFF that includes metadata and OME-Zarr for spatial transcriptomics.

Reviewer: I propose to replace 'OME-Zarr for spatial transcriptomics' by 'OME-Zarr is a cloud-friendly image format designed to efficiently store and access very large, multi-dimensional bioimaging datasets (such as spatial transcriptomics) along with their metadata.'
